api/keyconjurer/responses.go

```diff
@@ -1,30 +1,119 @@
 package keyconjurer
 
-// Response is the generic structure of the lambda responses
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+)
+
+// Response is the generic structure of the lambda responses.
+//
+// The unmarshalling of this structure is particularly poorly handled due to backwards compatibility concerns.
 type Response struct {
 	Success bool
+	// DEPRECATED: Use ErrorData.Message in the Data field if you intend to communicate error messages to the user.
 	Message string
-	Data    interface{}
+	// Data is the data that will be shipped to the user.
+	// Because it is not possible to UnmarshalJSON to interface{}, you must use the GetPayload() or GetError() functions instead to interact with this when unmarshalling from JSON.
+	// This is a bad pattern but it's a quick fix.
+	Data interface{}
+}
+
+func (r *Response) UnmarshalJSON(b []byte) error {
+	var inner struct {
+		Success bool
+		Message string
+		Data    json.RawMessage
+	}
+
+	if err := json.Unmarshal(b, &inner); err != nil {
+		return err
+	}
+
+	r.Success = inner.Success
+	r.Message = inner.Message
+	r.Data = inner.Data
+	return nil
+}
```

Comment:
Can we elaborate on why the backwards compatibility makes unmarshalling difficult? I'm not sure i understand the need for some of these functions

Reply:
I'll' add an explanation inline and explain in our 1-1 later